R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+  }
+
+  join_keys_obj[[dataset_1]] <- value
+
+  # Iterate on all new values to create symmetrical pair
+  for (ds2 in names(value)) {
+    if (ds2 == dataset_1) next
+
+    keep_value <- join_keys_obj[[ds2]] %||% list()
+    new_value <- value[[ds2]]
+
+    if (checkmate::test_character(new_value, min.len = 1, names = "unnamed")) {
+      new_value <- setNames(new_value, new_value)
+    } else if (
+      checkmate::test_character(new_value, min.len = 1)
+    ) {
+      # Invert key
+      new_value <- setNames(names(new_value), new_value)
+    }
+
+    keep_value[[dataset_1]] <- new_value
+
+    # Assign symmetrical
+    join_keys_obj[[ds2]] <- keep_value
+  }
+
+  # Remove NULL or empty keys
+  empty_ix <- vapply(
+    join_keys_obj,
+    function(.x) is.null(.x) || length(.x) == 0,
+    logical(1)
+  )
+  preserve_attr <- attributes(join_keys_obj)[!names(attributes(join_keys_obj)) %in% "names"]
+  join_keys_obj <- join_keys_obj[!empty_ix]
+  attributes(join_keys_obj) <- modifyList(attributes(join_keys_obj), preserve_attr)
+
+  #
+  # restore class
+  class(join_keys_obj) <- c("join_keys", "list")
+  join_keys_obj
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @keywords internal
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `join_keys` objects into the current `join_keys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`join_keys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `join_keys` objects or single `join_keys` object
+#'
+#' @return a new `join_keys` object with the resulting merge.
+#'
+#' @keywords internal
+merge_join_keys.join_keys <- function(join_keys_obj, new_join_keys) {
+  if (
+    checkmate::test_class(new_join_keys, "join_key_set") ||
+      checkmate::test_class(new_join_keys, "join_keys")
+  ) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  lapply(new_join_keys, assert_join_keys_alike)
+
+  if (checkmate::test_list(new_join_keys, types = "join_key_set")) {
+    jk_temp <- new_join_keys()
+    join_keys(jk_temp) <- new_join_keys
+    new_join_keys <- list(jk_temp)
+  }
+
+  checkmate::assert_list(new_join_keys, types = c("join_keys"), min.len = 1)
+
+  for (el in new_join_keys) {
+    join_keys_obj <- utils::modifyList(join_keys_obj, el)
+  }
+
+  logger::log_trace("join_keys keys merged.")
+  return(join_keys_obj)
+}
+
+# S3 methods have to be exported, otherwise `.S3method` needs to be used
+.S3method("merge_join_keys", "teal_data", merge_join_keys.default)
+.S3method("merge_join_keys", "join_keys", merge_join_keys.join_keys)
+
+#' Length of `join_keys` object.
+#' @inheritParams base::length
+#' @export
+length.join_keys <- function(x) {
```

Comment:
How is this different from using `length` directly? :thinking:

Reply:
To avoid leftovers inside `join_keys`. It's an edge case at the current version though (as empty elements are removed).

We can probably remove it for most use cases (I'll bring this to a call about JoinKeys)

``` r
length(list(ds1 = NULL))
#> [1] 1
length(list(ds1 = list()))
#> [1] 1
```